paddle/operators/listen_and_serv_op.cc

```diff
@@ -0,0 +1,207 @@
+    if (!rpc_service_) {
+      std::string endpoint = Attr<std::string>("endpoint");
+      rpc_service_.reset(new detail::AsyncGRPCServer(endpoint));
+      server_thread_.reset(new std::thread(RunServer, rpc_service_));
+    }
+  }
+
+  void Stop() override {
+    detail::MessageWithName term_msg;
+    term_msg.first = LISTEN_TERMINATE_MESSAGE;
+    rpc_service_->Push(term_msg);
+    rpc_service_->ShutDown();
+    server_thread_->join();
+  }
+
+  std::string GetGradVarNameForTrainer(const std::string &varname) const {
+    if (grads_counter_.find(varname) == grads_counter_.end()) {
+      grads_counter_[varname] = 0;
+    }
+    return string::Sprintf("%s.trainer_%d", varname, grads_counter_[varname]++);
+  }
+
+  void Run(const framework::Scope &scope,
+           const platform::Place &dev_place) const override {
+    platform::DeviceContextPool &pool = platform::DeviceContextPool::Instance();
+    auto &dev_ctx = *pool.Get(dev_place);
+    framework::Scope &recv_scope = scope.NewScope();
+
+    // FIXME(Yancey1989): initialize rpc server with lazy mode.
+    rpc_service_->SetScope(&recv_scope);
+    rpc_service_->SetDevCtx(&dev_ctx);
+    auto param_list = Attr<std::vector<std::string>>("ParamList");
+    auto grad_list = Attr<std::vector<std::string>>("GradList");
+    auto fan_in = Attr<int>("Fanin");
+
+    auto *block = Attr<framework::BlockDesc *>(kOptimizeBlock);
+    auto *program = block->Program();
+    framework::Executor executor(dev_place);
+
+    // TODO(typhoonzero): change this to a while_op for every cluster-batch.
+    bool exit_flag = false;
+    while (!exit_flag) {
+      // Get from multiple trainers, we don't care about the order in which
+      // the gradients arrives, just add suffix 0~n and merge the gradient.
+      rpc_service_->SetCond(0);
+      size_t recv_var_cnt = 0;
+      int batch_barrier = 0;
+      while (batch_barrier != fan_in) {
+        const detail::MessageWithName &v = rpc_service_->Get();
+        auto grad_var_name = v.first;
+        if (grad_var_name == LISTEN_TERMINATE_MESSAGE) {
+          LOG(INFO) << "received terminate message and exit";
+          exit_flag = true;
+          break;
+        } else if (grad_var_name == BATCH_BARRIER_MESSAGE) {
+          VLOG(3) << "recv batch barrier message";
+          batch_barrier++;
+          continue;
+        } else {
+          // receive a variable
+          recv_var_cnt++;
+          auto it =
+              std::find(grad_list.begin(), grad_list.end(), grad_var_name);
+          std::string param_var_name;
+          if (it != grad_list.end()) {
+            param_var_name = param_list[it - grad_list.begin()];
+          } else {
+            LOG(ERROR) << "grad has no paired param:" << grad_var_name;
+          }
+          VLOG(3) << "received grad: " << grad_var_name
+                  << " updating param: " << param_var_name;
+
+          if (fan_in > 1) {
+            grad_var_name = this->GetGradVarNameForTrainer(grad_var_name);
```

Comment:
Not a blocker for merging this PR:

I think the transpiler should handle this naming logic, rather than baked in the listen and serve operator.
If the trainspiler handles this naming logic, we don't need the fan_in variable anymore too: the operator can just wait all the vars being received, and run the optimize block. In this way the retry could be handled better as well: any retry will just override the previous sent variable.

Reply:
I think name the `ProgramDesc` maybe better than name the variable.
- Name the variable need to hold the relationship between original and the expansion.It should be held in trainer and pserver also.

What we want to do is distinguish where the variables come from, so an attribute of send_op to can do this.

And `fan_in` is useful when scaling down.
- The Kubernetes Job is a set, and we don't know which trainer will be killed when we set Job's parallelism.

Reply:
> The Kubernetes Job is a set, and we don't know which trainer will be killed when we set smaller Job's parallelism.fan_in is a count and doesn't care who it is.

We can achieve scaling down without fan_in as well. If we don't change fan_in, then the trainer will block because of the scale down. If we change fan_in, that means a new ProgramDesc is running on the pserver, we can do whatever we want in the new ProgramDesc, not just limited to changing fan_in only.

> I think name the ProgramDesc maybe better than name the variable.

Sorry what do you mean by "name the ProgramDesc"? And why naming the variable needs more memory?

Reply:
Thanks @helinwang 

> I think the transpiler should handle this naming logic (e.g., the gradient for variable W will be W.trainer.0, and W.trainer.1 on trainer 0 and trainer 1 respectively), rather than baked in the listen and serve operator.

That's exactly what I was thinking how to remove the executor `create_var` argument. I'll create an issue to do that in another PR.